src/SortingAlgorithms.jl

```diff
@@ -619,4 +640,54 @@ function sort!(v::AbstractVector, lo::Int, hi::Int, ::CombSortAlg, o::Ordering)
     return sort!(v, lo, hi, InsertionSort, o)
 end
 
+function sort!(v::AbstractVector, lo::Int, hi::Int, ::BitonicSortAlg, o::Ordering)
+    return bitonicsort!(view(v, lo:hi), o::Ordering)
+end
+
+function bitonicsort!(data, o::Ordering)
+    N = length(data)
+    for n in 1:ceil(Int, max(0, log2(N)))
+        bitonicfirststage!(data, Val(n), o::Ordering)
+        for m in n-1:-1:1
+            bitonicsecondstage!(data, Val(m), o::Ordering)
+        end
+    end
+    return data
+end
+
+function bitonicfirststage!(v, ::Val{Gap}, o::Ordering) where Gap
+    N = length(v)
+    @inbounds begin
+        gap = 1 << Gap
+        g2 = gap >> 1
+        for i in 0:gap:N-1
+            firstj = max(0, i + gap - N)
+            for j in firstj:(g2-1)
+                ia = i + j
+                ib = i + gap - j - 1
+                a, b = v[ia + 1], v[ib + 1]
+                v[ia+1], v[ib+1] = lt(o, b, a) ? (b, a) : (a, b)
+            end
+        end
+    end
+end
+
+function bitonicsecondstage!(v, ::Val{Gap}, o::Ordering) where Gap
+    N = length(v)
+    @inbounds begin
+        for n in Gap:-1:1
+            gap = 1 << n
+            for i in 0:gap:N-1
+                lastj = min(N - 1, N - (i + gap >> 1 + 1))
+                for j in 0:lastj
```

Comment:
I'm still trying to understand it all, because in the sorting network literature it's all about the delay, or depth... On the wikipedia page, though, it clearly states the network has `n log(n)^2` comparators, so I guess that should be the complexity. Not `n log(n)` but not too shabby, and the main attraction is in parallelism anyways.

By eyeballing the diagram, or unrolling these two stages, we have this:
```
first_stage(data, 1)

first_stage(data, 2)
second_stage(data, 1)

first_stage(data, 3)
second_stage(data, 2)
second_stage(data, 1)

first_stage(data, 4)
second_stage(data, 3)
second_stage(data, 2)
second_stage(data, 1)
```
So it's `log(n)` number of these larger blocks, but they increase linearly, making it `O(log(n)^2)` function calls. Each of these are linear, so `O(n log(n)^2)`. Do you agree? Notice it's linear because although it's two for-loops, the intervals are such that it's actually a linear traversal. We might even `reshape` the input to implement this.

Reply:
I agree that that is what the runtime _should_ be, but I think there is an error in the implementation that results in much larger runtimes.

Reply:
Oops, you're right! There was an extra outer for-loop for the second stage... Hopefully that helps a bit, but I'm still skeptical this is as good as it gets. I mean, I'm not even sure it will really be worth it, but I'm not sure we're at the best for this algorithm yet.